lib/SILGen/SILGenType.cpp

```diff
@@ -415,8 +415,19 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
     // If it's not an accessor, just look for the witness.
     if (!reqAccessor) {
       if (auto witness = asDerived().getWitness(reqDecl)) {
+        auto newDecl = requirementRef.withDecl(witness.getDecl());
+        // Only import C++ methods as foreign. If the following
+        // Objective-C function is imported as foreign:
+        //   () -> String
+        // It will be imported as the following type:
+        //   () -> NSString
+        // But the first is correct, so make sure we don't mark this witness
+        // as foreign.
+        if (dyn_cast_or_null<clang::CXXMethodDecl>(
+                witness.getDecl()->getClangDecl()))
+          newDecl = newDecl.asForeign();
```

Comment:
Why are C++ methods different than Objective-C methods in this regard?

Reply:
Basically, if an Objective-C method is marked as foreign it will be imported using the Objective-C type and not the Swift type. For example, it might be imported as `() -> NSString` instead of `() -> String` (which is what we want).